### DISCUSSION TITLE
Rethinking casts for the nth time

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#8675] [Ideas]
Here is what I would like to propose:

- `checked_cast(exp)` - replaced by `checked_cast(exp as type)`. The problem with `checked_cast` without an explicit type is that it wrecks too much havoc on type inference. It is simply to brittle. The type inference cannot also make the right guess. Worse, changes to type inference can break existing code. Its better to be explicit and safe. Moreover, `checked_cast` is almost never used any more-- since Java constructor/method resolution has changed. The std lib only has 6 uses of `checked_cast` and probably they can be refactored. 
- `checked_ecast` - GONE. Replaced by sub-effecting. And if sub-effecting is not enough then you can eta expand, i.e. introduce a lambda.
- `unchecked_cast` - kept as before.

**Comment by JonathanStarup** [comment]
we should use `checked_cast(ex : type)` to avoid niche unique syntax

**Reply by magnus-madsen** [reply]
Agree. If it is parseable.
